base/iterators.jl

```diff
@@ -1431,47 +1431,64 @@ only(x::NamedTuple) = throw(
 Base.intersect(a::ProductIterator, b::ProductIterator) = ProductIterator(intersect.(a.iterators, b.iterators))
 
 """
-    IterableClosure(f)
+    Iterator(f, initialstate)
 
-An iterable object that holds a function with no parameters, `f()`, that returns an optional `Union{Some{X},Nothing}`.
+Iterable object with iterations defined by the function `f(state)`. The function takes the current state at each iteration, and follows the same
+rules as `iterate`. It must return either `(newvalue, newstate)` or `nothing`, in which case the iterator ends.
 
-`f()` is supposed to be a closure, or thunk. The iterator will produce the values from the Some objects until `nothing` is returned, signaling the end of the iterator.
+See also: [`iterate`](@ref)
 
-## Examples
+# Examples
 
 ```jldoctest
-julia> function myrange(a, b)
-           state = a
-           IterableClosure() do
-               if state <= b
-                   prevstate, state = state, state + 1
-                   Some(prevstate)
-               else
-                   nothing
-               end
+julia> fib = Iterator((1,1)) do (a,b)
+           b, (a+b, a)
+       end;
+
+julia> reduce(hcat, Iterators.take(fib, 7))
+1×7 Matrix{Int64}:
+ 1  1  2  3  5  8  13
+
+julia> frac(c, z=0.0im) = Iterator{ComplexF64}((c, z)) do (c, z)
+           if real(z * z') < 4
+               z, (c, z^2 + c)
+           else
+               nothing
            end
-       end
-myrange (generic function with 1 method)
-
-julia> for x in myrange(3,5)
-           println(x)
-       end
-3
-4
-5
+       end;
+
+julia> [length(collect(frac(-0.835-0.2321im, (k+j*im)/6))) for j in -4:4, k in -8:8]
```

Comment:
Nice julia set! There's no need for `collect`, `[count(Returns(true), frac(...)) ...]` should work without `O(n*m)` allocations of unknown size :)

Reply:
Cool, I'll test it later to make sure

Reply:
Thanks, that's really much better. TIL!

Reply:
You're welcome! Keeping CI/doc generation time down is always a plus, imo.